Add PrestoSparkExecutionContext for capturing different runtime info for logging

pgupta2 opened the pull request:
## Description
PrestoSparkExecutionContext is being added to capture runtime information of 
PoS query which is currently not available like optimizations triggered at
runtime, auto tuned configs etc. QueryInfo is extended to include this context
object which can be used for quick analysis and for logging purposes. 

## Motivation and Context
Helps in capturing runtime metadata for PoS queries  

## Test Plan
Manually ran a query that triggers the auto tuning optimizations and verified that 
PrestoSparkExecutionContext information is getting logged properly

```
== NO RELEASE NOTE ==
```

agrawaldevesh commented on presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkPhysicalResourceCalculator.java:
```diff
@@ -63,7 +64,7 @@ public PhysicalResourceSettings calculate(PlanNode plan, PrestoSparkSourceStatsC
         int hashPartitionCount = getHashPartitionCount(session);
         OptionalInt maxExecutorCount = OptionalInt.empty();
-        PhysicalResourceSettings defaultResourceSettings = new PhysicalResourceSettings(hashPartitionCount, maxExecutorCount);
+        PhysicalResourceSettings defaultResourceSettings = new PhysicalResourceSettings(hashPartitionCount, maxExecutorCount, ImmutableSet.of());
```
Do you want to call this extraClientTags ?

pgupta2 replied:
Are you suggesting that we should rename `physicalResourceOptimizationTags` to `extraClientTags` or asking if we can use a better name instead of `extraClientTags` inside QueryInfo? Sorry, bit confused about the comment.

singcha replied:
Can we call it something like - `runtimeTags`, `serverTags`. I believe the naming of `clientTags` represents tags passed by clients ?

agrawaldevesh commented:
Hi @tdcmeehan, I was wondering if you can help review this or suggest who might be a good person ? Thanks.

agrawaldevesh commented:
@pgupta2 , can you change the description and title to say "Adding EXTRA client tags" ?

shrinidhijoshi reviewed (changes requested):
@pgupta2 Can we just add a new method `addTags()`  to `Session` class ? This would help collect tags that come up after query execution starts.

singcha commented on presto-main/src/main/java/com/facebook/presto/event/QueryMonitor.java:
```diff
@@ -410,15 +412,15 @@ private QueryStatistics createQueryStatistics(QueryInfo queryInfo)
     }
 
-    private QueryContext createQueryContext(SessionRepresentation session, Optional<ResourceGroupId> resourceGroup)
+    private QueryContext createQueryContext(SessionRepresentation session, Optional<ResourceGroupId> resourceGroup, Set<String> extraClientTags)
```
Can we create overloaded method with default value for extraClientTags as `ImmutableSet.of()`. This will avoid changing all the other places where we are simply passing empty maps

tdcmeehan commented:
After poking around at this a bit, it seems we're doing this so we can log the physical resource settings used.  That's because it seems we need to know the exact physical resource settings actually picked, since the session properties themselves don't tell you that.  My question is, if that's the case, why not just add the `PhysicalResourceSettings` to `QueryInfo` (OK to be null for streaming execution)?

pgupta2 commented:
> @pgupta2 Can we just add a new method `addTags()` to `Session` class ? This would help collect tags that come up after query execution starts.

The only way to mutate `Session` object is via `SessionBuilder` and we cannot instantiate SessionBuilder once the transaction starts due to the check here https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/Session.java#L619.  Also, `clientTags` is immutable Set and the variable is final. So, you cannot assign it to new Set or add anything new in the existing Set.

So, either we need to remove the transaction check which I think is dangerous or make `clientTags` a mutable Set in `Session` itself. My first thought was to make the `clientTags` mutable since the change seemed harmless to me but was not sure if this would be an accepted solution.

pgupta2 commented:
> After poking around at this a bit, it seems we're doing this so we can log the physical resource settings used. That's because it seems we need to know the exact physical resource settings actually picked, since the session properties themselves don't tell you that. My question is, if that's the case, why not just add the `PhysicalResourceSettings` to `QueryInfo` (OK to be null for streaming execution)?

@tdcmeehan : PhysicalResourceSettings is just one example of optimization triggering during query planning. There could be other cases as well where we would want to add extra tags to the query. For example: We would like to tag queries triggering specific AQE optimization in future. The idea here is to add the basic support of adding additional tags to the query during query compilation/runtime. This is the reason why `QueryInfo` takes in a Set of `extraTags` instead of `PhysicalResourceSettings`. With the current logic, we just need to collect the tags and make them part of `QueryInfo` and these will be logged along with Session.clientTags as part of EventListener events.

tdcmeehan commented:
Can such additional optimizations also be added to the `QueryInfo` when needed?  My concern is "extra client tags" is just too generic and opaque for someone trying to understand the behavior of their query, when it seems we're interested in actually very specific and precise things: what physical resource settings were chosen, what AQE optimizations were applied, etc.

pgupta2 commented:
> Can such additional optimizations also be added to the `QueryInfo` when needed? My concern is "extra client tags" is just too generic and opaque for someone trying to understand the behavior of their query, when it seems we're interested in actually very specific and precise things: what physical resource settings were chosen, what AQE optimizations were applied, etc.

We can do it but wouldn't it become ugly overtime? We might end up adding 10s of tags from different places and extending `QueryInfo` every time would be too much work. In the current case, we had the option of encapsulating the `tags` inside `PhysicalResourceSettings` class but this might not be possible in all the cases. In some cases, we might just want to add a tag if a certain condition is invoked and log that tag. In that case, we can have a Set<String> where we can keep adding tags and finally pass the whole thing to `QueryInfo`.

`extraTags` in QueryInfo is supposed to be generic in nature. It does not care who adds the tags. It only allows a mechanism to pass the Set<String>  and append it to clientTags coming from `Session` internally. The responsibility of adding the tag will be at the optimization level. For example, in this case, `PrestoSparkQueryExecutionFactory` picks the tags from `PhysicalResourceSettings` and pass it to `QueryInfo`. In future, we can pick tags coming from different sources, concat them and then pass it to `QueryInfo`.  

Also, we won't have to pass 'null' values for things that does not apply to Presto like `PhysicalResourceSettings`. This should keep the code clean. Right?

tdcmeehan commented:
Do we need to log this information as tags, or as informational fields?  For example, these tags might be just represented as `PhysicalResourceSettings`.  That's very familiar in spirit to existing data structures on the `QueryInfo` class such as `StatsAndCosts`.  If you're worried about cluttering up the `QueryInfo` class, just embed it as a new type, perhaps `SparkExecutionContext`, that contains important information that governed the Spark runtime.  It being null or empty is expected--`QueryInfo`'s sole purpose is to serialize to JSON, and empty portions are just omitted from that JSON.

Think of it this way.  When I look at a `QueryInfo`, I really want to make sure it's immediately completely intelligible.  I'm most likely looking at it because something's not working right.  When I see things like QueryType, QueryState, MemoryPoolId, these are all concepts that are directly mappable to real things I can describe in Presto.  There's no indirection.  A tag is a level of indirection--we can have multiple tags all together, and I need to interpret their meaning from the name (most likely by looking at the code, in this case finding the string constant by doing a code search).  And it's different from all the rest of the information in `QueryInfo`--for some reason, we'll log some information as JSON data structures that are self-describing, but some other information is expressed as tags.  To me, introducing two ways of describing things (tags and self describing data structures) it is more cluttered than sticking to just the latter.

To me, this:

```json
{
    physicalResourceSettings: {
        maxExecutorCount: 1000,
        hashPartitionCount: 500
    }
}
```

is just more readable and instantly understable than:

```json
{
    extraClientTags: [AUTO_TUNED_HASH_PARTITION_COUNT_TO_1000, AUTO_TUNED_MAX_EXECUTORS_COUNT_TO_500]
}
```

... when you start adding information which is not similar.  I need to know that for auto tuned hash partition counts specifically, I need to look at `extraClientTags` (and understand that this actually doesn't relate in any way to the client).  The more tags with different information that's added, the more difficult I think it will become to understand quickly.  If you want to know specifically that something was auto-tuned, we can add a boolean to indicate that.

Regarding the future potential need of adding lots of generic information, can we just say YAGNI and address that when the need arises?  For AQE, I can't imagine we couldn't create a datastructure that expresses this in a similar way that existing data structures on `QueryInfo` already do.

pgupta2 commented:
@tdcmeehan : Thanks for the detailed comment and explaining your thought process. I did not understood your comments initially and was looking into this as a way of extending clientTags for capturing runtime information as tags. But your suggestion is more around capturing the actual information itself which makes more sense.

I have updated this PR and added a `PrestoSparkExecutionContext` class which can be used for capturing any runtime state like auto tuned configs, runtime optimizations etc. and made it a part of QueryInfo which can then be logged as a JSON.

vermapratyush commented on presto-spi/src/main/java/com/facebook/presto/spi/PrestoSparkExecutionContext.java:
```diff
@@ -0,0 +1,95 @@
+        return physicalResourceSettings;
+    }
+
+    public static PrestoSparkExecutionContext create(
```
I suppose we will have more such static function to add new attributes to `PrestoSparkExecutionContext`.
In that case, should this function be called `.withResourceSettings`, like a builder pattern.

pgupta2 replied:
@vermapratyush : I was thinking more in terms of just extending the method signature with new attributes and then create and return an instance of PrestoSparkExecutionContext. Something similar to `QueryStats.create()`

tdcmeehan commented on presto-main/src/test/java/com/facebook/presto/execution/TestQueryInfo.java:
```diff
@@ -201,6 +201,7 @@ private static QueryInfo createQueryInfo()
                         new PlanNodeCanonicalInfo("a", ImmutableList.of()))),
-                ImmutableMap.of());
+                ImmutableMap.of(),
+                Optional.empty());
```
Should we enhance this test to add the `PrestoSparkExecutionContext` to prove it doesn't have issues serializing/deserializing to JSON?

tdcmeehan commented on presto-spi/src/main/java/com/facebook/presto/spi/PrestoSparkExecutionContext.java:
```diff
@@ -0,0 +1,95 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi;
```
Let's move this to a new package `com.facebook.presto.spi.prestospark` or something similar.  If we create that new package, I would also recommend putting the physical resource settings there so we don't need the inner class below, we can just directly annotate the original class.

tdcmeehan reviewed (commented):
LGTM % couple nits

pgupta2 commented:
@tdcmeehan : Addressed your comments. Please review again. Thanks.

pgupta2 commented:
@tdcmeehan : Gentle ping for review.

tdcmeehan reviewed (approved):
LGTM--thanks!

shrinidhijoshi reviewed (approved):
LGTM.  This was a good suggestion @tdcmeehan. Thanks for making the changes @pgupta2 .
